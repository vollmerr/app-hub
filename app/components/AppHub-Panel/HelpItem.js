import React from 'react';
import styled from 'styled-components';

import Link from 'components/Link';
import theme from 'utils/theme';

export const Item = styled.li`
  list-style-type: none;
  margin: 0 -20px;

  &:hover {
    background: ${theme.neutralLighter};
  }
`;

export const StlyedLink = styled(Link) `
  color: ${theme.white};
  display: block;
  padding: 15px;

  &:hover,
  &:active,
  &:focus {
    color: ${theme.neutralDark};
  }
`;

class HelpItem extends React.PureComponent {
  render() {
    return (
      <Item>
        <StlyedLink {...this.props} />
      </Item>
    );
  }
}

export default HelpItem;
