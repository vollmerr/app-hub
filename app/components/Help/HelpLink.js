import styled from 'styled-components';

import theme from '../../utils/theme';

import Link from '../Link';


const HelpLink = styled(Link) `
  color: ${theme.themePrimary};

  &:hover {
    color: ${theme.themeDarker};
    text-decoration: underline;
  }
`;


export default HelpLink;
