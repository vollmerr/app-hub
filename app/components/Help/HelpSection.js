import React from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';

import theme from '../../utils/theme';


export const Content = styled.div`
  padding: ${theme.hub.padding}px;
  padding-top: 0;
`;


export const Line = styled.div`
  height: 1px;
  margin: auto;
  width: 90%;
  background: ${theme.neutralLight};
`;


class HelpSection extends React.PureComponent {
  render() {
    const { title, desc } = this.props;

    return (
      <div>
        <h2>{title}</h2>
        <Content>{desc}</Content>
        <Line />
      </div>
    );
  }
}


const { string, node } = PropTypes;

HelpSection.propTypes = {
  title: string.isRequired,
  desc: node.isRequired,
};


export default HelpSection;
