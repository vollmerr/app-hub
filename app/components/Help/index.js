import React from 'react';
import PropTypes from 'prop-types';

import LayoutSection from '../Layout/Section';

const Section = ({ title, desc }) => (
  <LayoutSection>
    <h2>{title}</h2>
    <div>{desc}</div>
  </LayoutSection>
);

class Help extends React.PureComponent {
  render() {
    const { title, sections } = this.props;

    return (
      <div>
        <LayoutSection>
          <h1>{title}</h1>
        </LayoutSection>
        {
          sections.map((section) => (
            <Section key={section.title} {...section} />
          ))
        }
      </div>
    );
  }
}


const { string, array, node } = PropTypes;


Section.propTypes = {
  title: string.isRequired,
  desc: node.isRequired,
};

Help.propTypes = {
  title: string.isRequired,
  sections: array.isRequired,
};

export default Help;
