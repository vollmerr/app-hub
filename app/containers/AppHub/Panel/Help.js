import React from 'react';

export default () => <div>in Help panel</div>;
