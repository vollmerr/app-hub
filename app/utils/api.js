import { delay } from 'redux-saga';
import { call, put } from 'redux-saga/effects';
import decode from 'jwt-decode';
import 'whatwg-fetch';

import { authUserDone } from '../containers/AppHub/actions';


export const DEFAULT_EXPIRE = 1000;
export const TOKEN = 'id_token';


/**
 * Determines if should fetch based off a cache time
 *
 * @param {Date} date             - date of cache
 * @param {Number} cacheSeconds   - number of seconds to cache
 *
 * @return {Bool}                 - if should fetch
 */
export function shouldFetch(date, cacheSeconds = 350) {
  return !date || (new Date(date).getTime() - (cacheSeconds * 1000)) > new Date().getTime();
}


/**
 * Sets a token for authenication in localStorage
 *
 * @param {String} token  - token to set
 */
export function setToken(token) {
  localStorage.setItem(TOKEN, token);
}


/**
 * Gets the authenication token from localStorage
 * @return {String}     - authenication token
 */
export function getToken() {
  return localStorage.getItem(TOKEN);
}


/**
 * Clears the authenication token from localStorage
 */
export function clearToken() {
  localStorage.removeItem(TOKEN);
}


/**
 * Helper function to valdate that token exists and is not expired
 *
 * @param {object} token  - jwt token
 *
 * @return {bool}         - valid ? true : false
 */
export function validToken(token) {
  try {
    const { exp } = decode(token);
    const now = new Date().getTime().toString().substring(0, 10);

    return Number(exp) > Number(now);
  } catch (e) {
    return false;
  }
}


/**
 * Helper function to put jwt token into global redux store
 *
 * @param {String} token    - jwt token
 *
 * @return {Number}         - exipration time
 */
export function* putToken(token) {
  try {
    const {
      sub: sam,
      exp,
      iat,
      ...rest
    } = decode(token);

    const expire = (exp - iat) * 1000; // convert from ms to s

    yield put(authUserDone({ sam, expire, ...rest }));
    return expire;
  } catch (error) {
    throw new Error(error);
  }
}


/**
 * Authenticates a user
 * Checks local storage for token first
 *
 * @param {Object} action   - redux action to dispatch (must have 'type')
 * @param {Number} maxTries - max tries to re attempt authenicating on fail
 */
export function* authenticate(action, maxTries = 3) {
  let expire = DEFAULT_EXPIRE; // try again in 1s by default
  let tries = 0;
  let error = new Error('Failed to authenticate');
  while (tries < maxTries) { // eslint-disable-line
    const token = getToken();
    // if valid local token use that, otherwise get new token
    if (validToken(token)) {
      // update user in global appHub state
      expire = yield call(putToken, token);
      // reset max tries
      tries = 0;
    } else {
      try {
        const options = {
          method: 'get',
          credentials: 'include',
        };
        const { id_token } = yield call(request, API.JWT, options); // eslint-disable-line
        // set into local storage for future authentication caching
        setToken(id_token);
        // update user in global appHub state
        expire = yield call(putToken, id_token);
        // reset max tries
        tries = 0;
      } catch (err) {
        // increment max tries
        tries += 1;
        error = err;
      }
      // max number of tires, set error
      if (tries === maxTries) {
        yield put(authUserDone(error));
      }
    }
    // authenicate again when token has expired
    yield call(delay, expire);
  }
}


/**
 * Parses the JSON returned by a network request
 *
 * @param  {object} response A response from a network request
 *
 * @return {object}          The parsed JSON from the request
 */
export function parseJSON(response) {
  if (response.status === 204 || response.status === 205) {
    return null;
  }
  return response.json();
}


/**
 * Checks if a network request came back fine, and throws an error if not
 *
 * @param  {object} response   A response from a network request
 *
 * @return {object|undefined} Returns either the response, or throws an error
 */
export function checkStatus(response) {
  if (response.status >= 200 && response.status < 300) {
    return response;
  }

  const error = new Error(response.statusText);
  error.response = response;
  throw error;
}


/**
 * Requests a URL, returning a promise
 *
 * @param  {string} url       The URL we want to request
 * @param  {object} [options] The options we want to pass to "fetch"
 *
 * @return {object}           The response data
 */
export function request(url, options) {
  return fetch(url, options)
    .then(checkStatus)
    .then(parseJSON);
}


/**
 * Requests a URL with jwt token in header for authorization, returning a promise
 * Waits until jwt token is set in localStorage before making call
 *
 * @param  {string} url       The URL we want to request
 * @param  {object} [options] The options we want to pass to "fetch"
 *
 * @return {object}           The response data
 */
export function requestWithToken(url, options = {}) {
  let timer = null;

  const requestIfToken = () => { // eslint-disable-line
    // get users jwt token form storage else they dont have one
    const token = localStorage.getItem('id_token') || null;
    // valid token, call api
    if (validToken(token)) {
      clearTimeout(timer);
      // config settings to send api
      const newOptions = {
        method: options.method || 'get',
        headers: {
          'Content-Type': 'application/json',
          Authorization: `Bearer ${token}`,
        },
        ...options,
      };
      // append optional body to request (POSTs)
      if (options.body) {
        newOptions.body = JSON.stringify(options.body);
      }

      return request(url, newOptions);
    }
    // wait again until token set
    timer = setTimeout(requestIfToken, 100);
  };

  return requestIfToken();
}
