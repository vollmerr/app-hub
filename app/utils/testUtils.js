import React from 'react';
import { shallow } from 'enzyme';
import 'jest-styled-components';

/**
 * Tests a styled-component for basic tests
 * such as rending correctly and adopting attributes
 * @param {Object} Component    - React component under test
 * @param {Object} extend       - React component the component under test extends /  inherits from
 */
export function testStyledComponent(Component, extend = null) {
  describe(`<${Component.displayName} />`, () => {
    let wrapper;
    beforeEach(() => {
      wrapper = shallow(<Component />);
    });

    it('should render correctly', () => {
      expect(wrapper).toMatchSnapshot();
    });

    it(`should render an <${Component.target}> tag`, () => {
      expect(wrapper.type()).toEqual(Component.target);
    });

    it('should have a className attribute', () => {
      expect(wrapper.prop('className')).toBeDefined();
    });

    it('should adopt a valid attribute', () => {
      const id = 'test';
      wrapper.setProps({ id });
      expect(wrapper.prop('id')).toEqual(id);
    });

    if (extend) {
      it(`should extend ${extend.name}`, () => {
        expect(wrapper.find(extend).exists()).toEqual(true);
      });
    } else {
      it('should not adopt an invalid attribute', () => {
        const attribute = 'test';
        wrapper.setProps({ attribute });
        expect(wrapper.prop('attribute')).toBeUndefined();
      });
    }
  });
}

/**
 * Tests a components for basic mapDispatchToProps information
 * @param {func} mapDispatchToProps   - components mapDispatchToProps function
 * @param {object} actions            - functions in mapDispatchToProps to test
 */
export function testMapDispatchToProps(mapDispatchToProps, actions) {
  const onAction = (action) => (
    `on${action.charAt(0).toUpperCase()}${action.slice(1)}`
  );

  let dispatch;
  let dispatchProps;
  beforeEach(() => {
    dispatch = jest.fn();
    dispatchProps = mapDispatchToProps(dispatch);
  });

  Object.keys(actions).forEach((action) => {
    describe(onAction(action), () => {
      it('should be injected', () => {
        expect(dispatchProps[onAction(action)]).toBeDefined();
      });

      it('should dispatch when called', () => {
        dispatchProps[onAction(action)]();
        expect(dispatch).toHaveBeenCalled();
      });
    });
  });
}
