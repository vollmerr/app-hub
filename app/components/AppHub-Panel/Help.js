/**
*
* Help
*
*/

import React from 'react';
import styled from 'styled-components';
import Wrapper from './Wrapper';

import HelpItem from './HelpItem';

export const links = [
  {
    key: 'conditionsOfUse',
    name: 'Conditions of Use',
    href: 'http://www.ca.gov/Use',
  },
  {
    key: 'privacyPolicy',
    name: 'Privacy Policy',
    href: 'http://www.ca.gov/Privacy',
  },
  {
    key: 'accessibility',
    name: 'Accessibility',
    href: 'http://www.ca.gov/Accessibility',
  },
  {
    key: 'contactUs',
    name: 'Contact Us',
    href: 'https://cdt.ca.gov/support/',
  },
];

export const Items = styled.ul`
  padding: 0;
  margin: 0;
`;

class Help extends React.PureComponent {
  render() {
    return (
      <Wrapper>
        <h2>Help</h2>
        <hr />
        <Items>
          {
            links.map((link) => <HelpItem key={link.key} href={link.href}>{link.name}</HelpItem>)
          }
        </Items>
      </Wrapper>
    );
  }
}

Help.propTypes = {

};

export default Help;
