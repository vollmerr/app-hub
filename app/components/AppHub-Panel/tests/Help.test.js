import React from 'react';
import { shallow } from 'enzyme';

import { testStyledComponent } from 'utils/testUtils';

import HelpItem from '../HelpItem';
import Help, { Items, links } from '../Help';

testStyledComponent(Items);


describe('<Help />', () => {
  let wrapper;
  beforeEach(() => {
    wrapper = shallow(<Help />);
  });

  it('should render correctly', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('should render a h2 header', () => {
    expect(wrapper.find('h2').length).toEqual(1);
  });

  it('should render all the links', () => {
    expect(wrapper.find(HelpItem).length).toEqual(links.length);
  });
});
