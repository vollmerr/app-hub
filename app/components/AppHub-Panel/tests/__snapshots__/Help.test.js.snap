// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Help /> should render correctly 1`] = `
<Wrapper__Wrapper>
  <h2>
    Help
  </h2>
  <hr />
  <Help__Items>
    <HelpItem
      href="http://www.ca.gov/Use"
      key="conditionsOfUse"
    >
      Conditions of Use
    </HelpItem>
    <HelpItem
      href="http://www.ca.gov/Privacy"
      key="privacyPolicy"
    >
      Privacy Policy
    </HelpItem>
    <HelpItem
      href="http://www.ca.gov/Accessibility"
      key="accessibility"
    >
      Accessibility
    </HelpItem>
    <HelpItem
      href="https://cdt.ca.gov/support/"
      key="contactUs"
    >
      Contact Us
    </HelpItem>
  </Help__Items>
</Wrapper__Wrapper>
`;

exports[`<Help__Items /> should render correctly 1`] = `
.c0 {
  padding: 0;
  margin: 0;
}

<ul
  className="c0"
/>
`;
